### PR TITLE
Move makefile tutorial from pre-lab 5 to in-lab 3

### PR DESCRIPTION
**Pull request by adeet1**
#41 

@50Wliu has endorsed this.

Compiling code with `make` is much easier and faster than having to manually compile several files via `clang++`. I thought it would be a good idea to move the makefile tutorial to in-lab 3, as that is much more lightweight than pre-lab 5. Since lab 5 is considered by many to be one of the more difficult labs, I thought that this would better balance the workload.

Lab 1 can teach students to compile manually with `clang++` (e.g. via the provided LifeCycle program). I have also provided a makefile for lab 2 that students can use (although at that point, they're not expected to understand any of the syntax since lab 2 is already intense enough).

### CODE REVIEW

**Comment by adeet1** [comment]
@50Wliu What do you think of these changes?

File: labs/lab02/index.md [hunk 1]
```diff
@@ -201,13 +201,9 @@ Thus, it makes sense that we should delete all the elements we inserted (hint: d
 However, what else do we dynamically allocate that we need to delete?
 
 #### Compiling ####
-When compiling your code, you must remember to compile all of your .cpp files in one line:
+To compile your code, simply run `make` in the terminal. We have provided a Makefile to make (pun intended) it easier for you to compile your code (instead of having to type `clang++ List.cpp ListItr.cpp ListNode.cpp ListTest.cpp` each time). You do not need to understand the syntax of the Makefile at this point, but if you're interested, you can go through the [makefile tutorial](../../tutorials/05-make/index.html), which will be required for in-lab 3.
```
Comment (50Wliu) [inline labs/lab02/index.md]: We may want this to be more prominent, especially if we use this to change ListTest to use `stoi`.

Prose-wise, I like it.

File: labs/lab02/index.md [hunk 2]
```diff
@@ -201,13 +201,9 @@ Thus, it makes sense that we should delete all the elements we inserted (hint: d
...
 
-```
-clang++ List.cpp ListItr.cpp ListNode.cpp ListTest.cpp
-```
-
-There are ways to compile these programs in pieces, but we will see this later in the semester.
+**Remark:** The provided source code will not compile out-of-the-box, so in order for `make` to run successfully, you will need to create the corresponding .cpp files for each of the .h files provided.
```
Comment (50Wliu) [inline labs/lab02/index.md]: Use Note as the other parts of the lab does.

File: labs/lab03/index.md [hunk 3]
```diff
@@ -34,16 +36,18 @@ Procedure
 
 1. Ensure your postfix calculator works on all the provided test cases
 2. Expand your test program to handle keyboard input
-3. Files to download: none (just your pre-lab source code)
-4. Files to submit: postfixCalculator.h, postfixCalculator.cpp, testPostfixCalc.cpp
+3. Complete the [makefile tutorial](../../tutorials/05-make/index.html).
```
Comment (50Wliu) [inline labs/lab03/index.md]: Probably should be first. Again, more important if we'll be using this to enable usage of stoi.

File: labs/lab03/index.md [hunk 4]
```diff
@@ -34,16 +36,18 @@ Procedure
...
 ### Post-lab ###
 
 1. Implement a stack class
 2. Modify your postfix calculator to use your stack rather than the STL stack
 3. Describe any difficulties you encountered getting your code working and what you did to solve them
-4. Files to download: none (just your in-lab source code)
-5. Files to submit: stack.h, stack.cpp, postfixCalculator.h, postfixCalculator.cpp, testPostfixCalc.cpp, difficulties.txt - You may submit additional stack/list files as well, if you want
+4. Your code must compile with `make`!
+5. Files to download: none (just your in-lab source code)
+6. Files to submit: stack.h, stack.cpp, postfixCalculator.h, postfixCalculator.cpp, testPostfixCalc.cpp, difficulties.txt, Makefile, and any other .h and .cpp files you need
```
Comment (50Wliu) [inline labs/lab03/index.md]: :+1:

File: labs/lab03/index.md [hunk 5]
```diff
@@ -34,16 +36,18 @@ Procedure
...
+5. Files to submit: postfixCalculator.h, postfixCalculator.cpp, testPostfixCalc.cpp, Makefile-pizza
 
 ### Post-lab ###
 
 1. Implement a stack class
 2. Modify your postfix calculator to use your stack rather than the STL stack
 3. Describe any difficulties you encountered getting your code working and what you did to solve them
-4. Files to download: none (just your in-lab source code)
-5. Files to submit: stack.h, stack.cpp, postfixCalculator.h, postfixCalculator.cpp, testPostfixCalc.cpp, difficulties.txt - You may submit additional stack/list files as well, if you want
+4. Your code must compile with `make`!
```
Comment (50Wliu) [inline labs/lab03/index.md]: Nah, we can talk about this in the in-lab and warn them that all code must compile with Make from now on.

**Review by winstliu — COMMENTED** [review]
This is of course all dependent on @aaronbloomfield's approval

**Comment by 50Wliu** [comment]
The Makefile itself is outdated based on the changes I made to the Makefile tutorial last semester. Take a look at the new Makefile-pizza.

File: labs/lab02/index.md (hunk 1 above)
Comment (adeet1) [reply labs/lab02/index.md]: @50Wliu Do you mean that we should put the Compiling section somewhere at the top of the pre-lab instructions, like after the UML diagram?

**Comment by adeet1** [comment]
@50Wliu

**UPDATE:** I discussed this pull request with @aaronbloomfield during office hours today, and he mentioned that moving the makefile tutorial to lab 3 may not be the best course of action for the following reasons:
1) Students aren't yet completely familiar with the Unix environment at this point in the semester.
2) Makefile syntax is not user-friendly and error messages are cryptic, so it will be difficult for students to properly edit provided makefiles and/or implement their own makefiles from scratch.

From my understanding, we can keep the makefile tutorial in lab 5 (as is), but we could provide fully implemented makefiles for labs 2 and 4, because students need to submit very specific .h/.cpp files for these labs (which means it wouldn't even be necessary to edit them, they'd just have to run `make`).

However, providing a fully implemented makefile for lab 3 wouldn’t be a good idea, because the requirements for this lab are flexible and students would need to edit the provided makefile to suit their specific implementation (which would cause problem 2 above).

**Pull request closed without merging**